synapse/storage/databases/main/event_federation.py

```diff
@@ -714,96 +722,227 @@ def _get_auth_chain_difference_txn(
-
-        We use this function so that we can compare and see if someones current
-        depth at their current scrollback is within pagination range of the
-        insertion event. If the current depth is close to the depth of given
-        insertion event, we can trigger a backfill.
+        """
+        Get the insertion events we know about that we haven't backfilled yet
+        along with the aproximate depth (sorted by depth descending).
+
+        We use this function so that we can compare and see if someones
+        `current_depth` at their current scrollback is within pagination range
+        of the insertion event. If the `current_depth` is close to the depth
+        of the given insertion event, we can trigger a backfill.
+
+        We ignore insertion events that have a greater depth than our `current_depth`
+        as:
+            1. we don't really care about getting events that have happened
+               after our current position; and
+            2. by the nature of paginating and scrolling back, we have likely
+               previously tried and failed to backfill from that insertion event, so
+               to avoid getting "stuck" requesting the same backfill repeatedly
+               we drop those insertion event.
 
         Args:
             room_id: Room where we want to find the oldest events
+            current_depth: The depth at the users current scrollback position because
+                we only care about finding events older than the given
+                `current_depth` when scrolling and paginating backwards.
 
         Returns:
-            List of (event_id, depth) tuples
+            List of (event_id, depth) tuples (sorted by depth descending)
         """
 
         def get_insertion_event_backward_extremities_in_room_txn(
             txn: LoggingTransaction, room_id: str
         ) -> List[Tuple[str, int]]:
             sql = """
-                SELECT b.event_id, MAX(e.depth) FROM insertion_events as i
+                SELECT insertion_event_extremity.event_id, event.depth FROM insertion_events as insertion_event
                 /* We only want insertion events that are also marked as backwards extremities */
-                INNER JOIN insertion_event_extremities as b USING (event_id)
+                INNER JOIN insertion_event_extremities as insertion_event_extremity USING (event_id)
                 /* Get the depth of the insertion event from the events table */
-                INNER JOIN events AS e USING (event_id)
-                WHERE b.room_id = ?
-                GROUP BY b.event_id
+                INNER JOIN events AS event USING (event_id)
+                /**
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info USING (room_id, event_id)
```

Comment:
Error as seen in the failing CI, https://github.com/matrix-org/synapse/runs/8123978698?check_suite_focus=true#step:4:4915, https://github.com/matrix-org/synapse/runs/8123978792?check_suite_focus=true#step:4:18612

```
 psycopg2.errors.AmbiguousColumn: common column name "room_id" appears more than once in left table
```

Not sure why I can't reproduce locally 🤔

Reply:
Solved by specifying specific columns instead of `using`

```sql
LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info
ON
    failed_backfill_attempt_info.room_id = insertion_event_extremity.room_id
    AND failed_backfill_attempt_info.event_id = insertion_event_extremity.event_id
```